manifestservice/manifests/__init__.py

```diff
@@ -0,0 +1,259 @@
+from flask import Blueprint, request, Flask
+
+import json
+import flask
+import requests
+import ntpath
+from datetime import date, datetime
+from authutils.token.validate import (
+    current_token,
+    require_auth_header,
+    validate_request,
+    set_current_token
+)
+
+blueprint = Blueprint("manifests", __name__)
+
+import boto3
+session = boto3.Session(region_name="us-east-1")
+s3 = session.resource("s3")
+
+def get_folder_name_from_token(user_info):
+    """
+    Returns the name of the user's manifest folder (their "prefix").
+    It takes a "user_info" dict, which is the response that Fence returns at /user/user
+    The convention we'll use here is that a user's folder name will be "user-x" where x is 
+    their ID (integer).
+
+    According to the revproxy's helpers.js, it looks like the user_id is stored in a variable called "sub". Hm. 
+    """
+    return "user-" + str(user_info["sub"])
+
+def does_the_user_have_read_access_on_at_least_one_project(current_token):
+    """
+    Returns True if the user has both read and read-storage access on at least one project, 
+    False otherwise.
+    """
+    privileges = []
+    try:
+        project_access_dict = current_token.get("context").get("user").get("projects")
+        privileges = list(project_access_dict.values())
+    except Exception:
+        return False
+    
+    if len(privileges) == 0:
+        return False
+
+    for auth_set in privileges:
+        if "read" in auth_set and "read-storage" in auth_set:
+            return True
+
+    return False
+
+def is_valid_manifest(manifest_json):
+    """
+    Returns (True, "") if the manifest.json is a list of the form [{'k' : v}, ...], 
+    where valid keys are object_id and subject_id.
+    Otherwise, returns (False, error_msg)
+    """
+    valid_keys = set(["object_id" , "subject_id"])
+    error_msg = "Manifest format is invalid. Please POST a list of key-value pairs, like [{'k' : v}, ...] Valid keys are: " + " ".join(valid_keys)
+    if type(manifest_json) != list:
+        return False, error_msg
+
+    if len(manifest_json) == 0:
+        return True, ""
+
+    for record in manifest_json:
+        record_keys = record.keys()
+        if not set(record_keys).issubset(valid_keys):
```

Comment:
I think we actually want to reverse this. `valid_keys` should be a subset 
 of `record_keys`, so `record_keys` can have more stuff if it wants to

Reply:
in other words, we can add stuff ot the manifest in the future so long as we still have the object_id and subject_id